Add support for AES CBC and ECB mode in mbedTLS shim

sidcha opened the pull request:
Add support for AES CBC and ECB mode of operation in mbedTLS shim driver. Refactor mtls_session_setup() to allow multiple cipher modes to co-exist. Also update samples/drivers/crypto to include for ECB mode.

Both these modes are needed for an upcoming change: #19483

sidcha committed:
drivers/crypto: Add support for AES CBC mode in mbedTLS shim

Add support for AES CBC mode of operation in mbedTLS shim driver.
Refactor mtls_session_setup() to allow multiple cipher modes to
co-exist.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

tbursztyka commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -133,7 +176,8 @@ static int mtls_session_setup(struct device *dev, struct cipher_ctx *ctx,
 
-	if (mode != CRYPTO_CIPHER_MODE_CCM) {
+	if (mode != CRYPTO_CIPHER_MODE_CCM &&
+			mode != CRYPTO_CIPHER_MODE_CBC) {
```
indent under last '(' + 1 space

cbsiddharth replied:
Done

tbursztyka reviewed (approved):
code looks fine. 

Btw, did you got samples/drivers/crypto working?

cbsiddharth commented:
Yes.

sidcha committed:
samples: crypto: Refactor variables for clarity

Rename global variable ciphertext used in cbc_mode() to
cbc_ciphertext and move it closer to cbc_mode(). Also, move
global variable iv into cbc_mode() to be consistent with
other *_mode() methods.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

sidcha committed:
samples: crypto: Add AES ECB mode sample

Add sample for AES ECB mode with FIPS-197 ECB test vetors

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

sidcha committed:
drivers/crypto: Add support for AES ECB mode in mbedTLS shim

Add support for AES ECB mode of operation in mbedTLS shim driver.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

cbsiddharth commented:
@tbursztyka, I have added some more commits to this PR. Initially I was planning on opening another PR for ECB mode, but that PR would depend on this anyway. So just added those commits here.

Please review again.

chrta commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -44,12 +48,102 @@ struct mtls_shim_session mtls_sessions[CRYPTO_MAX_SESSION];
+{
+	int ret;
+	mbedtls_aes_context *ecb_ctx =
+		&((struct mtls_shim_session *)ctx->drv_sessn_state)->mtls_aes;
```
Maybe you could add a macro to do this: `((struct mtls_shim_session *)ctx->drv_sessn_state)` since this is done in a few places.

cbsiddharth replied:
something like this?
```
#define get_mtls_ctx(c, m) &((struct mtls_shim_session *)c->drv_sessn_state)->mtls_ ## m

mbedtls_aes_context *ctx = get_mtls_ctx(ctx, aes);
```

chrta replied:
Yes, that looks better (for me).

cbsiddharth replied:
Done.

sidcha committed:
drivers/crypto: Add macro to simplify access to ctx->drv_sessn_state

The conversion of struct cipher_ctx * to mtls_aes_context * happens in
a bunch of places. Add a macro MTLS_GET_CTX() to simplify this.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

cbsiddharth commented:
CI failure looks unrelated. Can someone please verify?

cbsiddharth commented:
@galak @nashif please review.
